src/framework/audio/internal/platform/jack/jackaudiodriver.cpp

```diff
@@ -0,0 +1,325 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
+ * MuseScore-CLA-applies
+ *
+ * MuseScore
+ * Music Composition & Notation
+ *
+ * Copyright (C) 2021 MuseScore BVBA and others
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include "jackaudiodriver.h"
+
+#include <fcntl.h>
+#include <unistd.h>
+#include <string.h>
+#include <math.h>
+#include <pthread.h>
+#include <thread> // Used by usleep
+#include <chrono> // Used by usleep
+
+#include "translation.h"
+#include "log.h"
+#include "runtime.h"
+
+static constexpr char DEFAULT_DEVICE_ID[] = "default";
+
+using namespace mu::audio;
+
+//namespace {
+struct JackData
+{
+    float* buffer = nullptr;
+    jack_client_t* jackDeviceHandle = nullptr;
+    unsigned long samples = 0;
+    int channels = 0;
+    std::vector<jack_port_t*> outputPorts;
+    pthread_t threadHandle = 0;
+    IAudioDriver::Callback callback;
+    void* userdata = nullptr;
+};
+
+static JackData* s_jackData{ nullptr };
+IAudioDriver::Spec s_format;
+
+static void* jackThread(void* jackParam)
+{
+    mu::runtime::setThreadName("audio_driver");
+    JackData* data = static_cast<JackData*>(jackParam);
+    jack_client_t* handle = data->jackDeviceHandle;
+
+    // complete jack register, setting the callback-process in this thread,
+    // which hopefully will be interrupted when data is ready from jack
+    jack_on_shutdown(handle, jack_cleanup_callback, 0);
+    jack_set_process_callback(handle, jack_process_callback, (void*)jackParam);
+
+    if (jack_activate(handle)) {
+        LOGE() << "cannot activate client";
+        return nullptr;
+    }
+
+    while (1) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+    }
```

Comment:
Maybe it's not done yet... probably there shouldn't be an endless empty loop with no way to exit it

Reply:
Thanks! fixed in a cleanup commit.